### PR TITLE
[SYSTEMML-998] Overview page as a design document for developers

### PR DESCRIPTION
**Pull request by niketanpansare**
Advantage of this approach:
- Will be helpful for a new contributor to learn SystemML internals
- Hyperinks to classes
- Code and documentation in the same file. Makes it easy to enforce "Making documentation part of a PR that affects user  APIs might be a good idea"
- Potentially could be exposed through systemml.apache.org
- Simplified documentation versioning

Note: this is just a developer documentation and is not a substitute for user guide/documentation.

@mboehm7 @fschueler @nakul02 @deroneriksson @dusenberrymw @lresende

### CODE REVIEW

**Comment by niketanpansare** [comment]
javadoc

**Comment by nakul02** [comment]
After talking with @niketanpansare and looking at the documentation for the [javadoc](https://maven.apache.org/plugins/maven-javadoc-plugin/examples/javadoc-resources.html) phase, I think the file `overview.html` should be in `src/main/javadoc/overview.html` and not where its currently put.

File: src/main/java/org/apache/sysml/overview.html [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+  <li>Command-line interface ( {@see org.apache.sysml.api.DMLScript} )</li>
+  <li>Convenient programmatic interface for Spark users ( {@see org.apache.sysml.api.mlcontext.MLContext} )</li>
+  <li>Java Machine Learning Connector API ( {@see org.apache.sysml.api.jmlc.Connection} )</li>
+</ul>
+
+{@see org.apache.sysml.parser.AParserWrapper} performs syntatic validation and 
+parses the input DML script using ANTLR into a 
+a hierarchy of {@see org.apache.sysml.parser.StatementBlock} and 
+{@see org.apache.sysml.parser.Statement}  as defined by control structures.
+Another important class of the language component is {  @see org.apache.sysml.parser.DMLTranslator}
```
Comment (nakul02) [inline src/main/java/org/apache/sysml/overview.html]: Extra space here - `{  @see org.apache.sysml.parser.DMLTranslator}` causes it to not be rendered correctly.

**Comment by nakul02** [comment]
@MadisonJMyers, @deroneriksson - do you think this content would be useful to have in the github wiki and link to it in the `README.md` ?

**Comment by nakul02** [comment]
and @niketanpansare - thank you!
this is very good work and a much needed effort in the direction of documentation that helps newcomer.

File: src/main/java/org/apache/sysml/overview.html (hunk 1 above)
Comment (niketanpansare) [inline src/main/java/org/apache/sysml/overview.html]: Good catch ... I will make the change :+1:

**Comment by deroneriksson** [comment]
@nakul02 When we have the javadocs deployed on the main website, this documentation will appear when a user goes to the javadocs, so I believe I would use this path to get to the overview.

**Comment by niketanpansare** [comment]
@deroneriksson @nakul02 I will move the overview.html to the directory that nakul recommended in my next commit :) ... Is there an umbrella JIRA for improving the documentation ?

**Comment by deroneriksson** [comment]
@niketanpansare there are probably 200 things needed for improving the documentation, so i'm not sure if a single umbrella JIRA handles it all.

**Comment by niketanpansare** [comment]
I understand. It just feels little awkward to push a PR without JIRA associated with it :(

**Comment by deroneriksson** [comment]
Please feel free to create a JIRA for it

**Comment by deroneriksson** [comment]
@niketanpansare Before merging please also verify that the overview renders correctly when 'mvn javadoc:javadoc ...' is run and also when 'mvn clean package -P distribution' generates the javadoc jar.

**Comment by niketanpansare** [comment]
Both  'mvn javadoc:javadoc ...' and 'mvn clean package -P distribution' are tested and works correctly on my setup after recent commit.

**Comment by deroneriksson** [comment]
@niketanpansare Great! I verified that both commands work correctly for me too with Java 8.

LGTM.

**Comment by niketanpansare** [comment]
To compile Python doc, you need to install Sphinx
```bash
pip install -U Sphinx
```

Then prepare docs:
```bash
cd src/main/python
sphinx-apidoc --full --force --output-dir docs systemml
```
Add following lines to src/main/python/docs/conf.py
```python
import sys
sys.path.append('/home/biuser/nike/python/apache/incubator-systemml/src/main/python')
autodoc_mock_imports = ['pyspark', 'py4j.java_gateway', 'pyspark.mllib.common', 'pyspark.context', 'pyspark.sql', 'pyspark.ml', 'pyspark.ml.feature']
```

Make the doc
```bash
cd docs
make singlehtml
or 
make html
```

Here is a sample document generated by above steps:
[python_systemml_documentation.pdf](https://github.com/apache/incubator-systemml/files/503900/python_systemml_documentation.pdf)

**Comment by niketanpansare** [comment]
Closed by the commit https://github.com/apache/incubator-systemml/commit/c7d990ce6f94b5ceae2ed37e62b66164cb15b8c8

**Pull request closed without merging**